File: haskell/isogram/src/Isogram.hs
module Isogram (isIsogram) where

import Data.Char (isLetter, toLower)
import Data.List (nub)

isIsogram :: String -> Bool
isIsogram string =
  let
    lowerCase = map toLower string
    letters = filter isLetter lowerCase
    unique = nub letters
  in
    length unique == length letters

File: erlang/isogram/src/isogram.erl
-module(isogram).
-export([is_isogram/1]).

is_isogram(Phrase) ->
  is_isogram(string:lowercase(Phrase), sets:new()).

is_isogram([], _) -> true;
is_isogram([$-|T], Chars) -> is_isogram(T, Chars);
is_isogram([$ |T], Chars) -> is_isogram(T, Chars);
is_isogram([H|T], Chars) ->
  case sets:is_element(H, Chars) of
    true -> false;
    false -> is_isogram(T, sets:add_element(H, Chars))
  end.
